Optimize Vector construction in a few more places

cdumez opened the pull request:
#### 91a785670c7f5984fd676b857e163acd53d87da8
<pre>
Optimize Vector construction in a few more places
<a href="https://bugs.webkit.org/show_bug.cgi?id=264712">https://bugs.webkit.org/show_bug.cgi?id=264712</a>

Reviewed by Brent Fulgham.

* Source/JavaScriptCore/b3/B3CheckSpecial.cpp:
(JSC::B3::CheckSpecial::hiddenBranch const):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::intlAvailableCalendars):
* Source/JavaScriptCore/wasm/WasmWorklist.cpp:
(JSC::Wasm::Worklist::Worklist):
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::copyDisjunction):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::domainToASCII):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::trackSizesForComputedStyle const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertContentAlignmentDataList):
* Source/WebKit/WebProcess/WebCoreSupport/WebPopupMenu.cpp:
(WebKit::WebPopupMenu::populateItems):

Canonical link: <a href="https://commits.webkit.org/270645@main">https://commits.webkit.org/270645@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e2ca521f1a1d7dc399f29e0dc615ea1978a8177

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25942 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4553 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27222 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28041 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23757 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26259 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6315 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1980 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23835 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26193 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3435 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22349 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28621 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3055 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23302 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29367 "Found 1 new test failure: http/tests/inspector/network/har/har-page.html (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22593 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23669 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23679 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27243 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25162 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1295 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32606 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4477 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7083 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6259 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3543 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3405 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham reviewed (approved):
I didn't realize `appendUsingFunctor` allowed you to avoid the preallocation. Very nice!

r=me

cdumez committed:
Optimize Vector construction in a few more places
https://bugs.webkit.org/show_bug.cgi?id=264712

Reviewed by Brent Fulgham.

* Source/JavaScriptCore/b3/B3CheckSpecial.cpp:
(JSC::B3::CheckSpecial::hiddenBranch const):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::intlAvailableCalendars):
* Source/JavaScriptCore/wasm/WasmWorklist.cpp:
(JSC::Wasm::Worklist::Worklist):
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::copyDisjunction):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::domainToASCII):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::trackSizesForComputedStyle const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertContentAlignmentDataList):
* Source/WebKit/WebProcess/WebCoreSupport/WebPopupMenu.cpp:
(WebKit::WebPopupMenu::populateItems):

Canonical link: https://commits.webkit.org/270645@main

cdumez commented on Source/JavaScriptCore/b3/B3CheckSpecial.cpp:
```diff
@@ -101,9 +101,9 @@ CheckSpecial::~CheckSpecial()
-    hiddenBranch.args.reserveInitialCapacity(m_numCheckArgs);
-    for (unsigned i = 0; i < m_numCheckArgs; ++i)
-        hiddenBranch.args.append(inst.args[i + 1]);
+    hiddenBranch.args.appendUsingFunctor(m_numCheckArgs, [&](size_t i) {
```
Oh, I think I understand. In this case, for example, we could have calls the Vector constructor and done an assignment for example? If so, I agree in general. It would have the benefit of calling reserveInitialCapacity() instead of reserveCapacity() internally. In this particular case though, args is a `Vector<Arg, 3>`. The fact that it has an inline buffer means that the move assignment operator actually may actually copying (assuming the inline buffer is used) so there would be a cost from that if we didn't use appendUsingFunctor.

I need to look at other cases in this patch too.

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/20431.
